# **Final Python Notebook 1**

### Author - Sanuli Gehara Jayasekara

### Student ID - 20220855 / w2053019

### Peer Reviewer - Krithiga Thachcanamorthi


In [222]:
# Import pandas Python library as pd to load and access coursework dataset
import pandas as pd

# Load the dataset as data frame
data_frame = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/CW Notebooks/datasets/5DATA002W.2 Coursework Dataset(25012025v6.0).csv")



Below code block is reused from the code reuse session 1. Also this is included in tutorial 1 to display the ifrst few rows in data frame

In [223]:
#displaying first five rows of the dataset
data_frame.head(10)


Patient_ID  Month_of_Birth   Age     Sex    Occupation T_Stage N_Stage  \
0      A0012              12  68.0  Female      Teaching      T1      N1   
1      A0013              12  50.0  Female       Medical      T2      N2   
2      A0014              11  58.0  Female   Engineering      T3      N3   
3      A0015               3  58.0  Female    Technology      T1      N1   
4      A0016               1  47.0  Female    Multimedia      T2      N1   
5      A0017               2  51.0  Female      Research      T1      N1   
6      A0018               5  51.0  Female  House Person      T1      N1   
7      A0019               4  40.0  Female     Reporting      T2      N1   
8      A0020               6  40.0  Female   Development      T4      N3   
9      A0021               4  69.0  Female  Construction      T4      N3   

  6th_Stage             Differentiated  Grade   A_Stage  Tumor_Size  \
0       IIA      Poorly differentiated      3  Regional         4.0   
1      IIIA  Moderately differentiated      2  Regional        35.0   
2      IIIC  Moderately differentiated      2  Regional        63.0   
3       IIA      Poorly differentiated      3  Regional        18.0   
4       IIB      Poorly differentiated      3  Regional        41.0   
5       IIA  Moderately differentiated      2  Regional        20.0   
6       IIA        Well differentiated      1  Regional         8.0   
7       IIB  Moderately differentiated      2  Regional        30.0   
8      IIIC      Poorly differentiated      3  Regional       103.0   
9      IIIC        Well differentiated      1   Distant        32.0   

  Estrogen_Status Progesterone_Status  Regional_Node_Examined  \
0        Positive            Positive                    24.0   
1        Positive            Positive                    14.0   
2        Positive            Positive                    14.0   
3        Positive            Positive                     2.0   
4        Positive            Positive                     3.0   
5        Positive            Positive                    18.0   
6        Positive            Positive                    11.0   
7        Positive            Positive                     9.0   
8        Positive            Positive                    20.0   
9        Positive            Positive                    21.0   

   Reginol_Node_Positive  Survival_Months Mortality_Status  
0                      1               60            Alive  
1                      5               62            Alive  
2                      7               75            Alive  
3                      1               84            Alive  
4                      1               50            Alive  
5                      2               89            Alive  
6                      1               54            Alive  
7                      1               14             Dead  
8                     18               70            Alive  
9                     12               92            Alive

These code blocks are taken from ocde reuse session 1. This helps to idnetify the dataset columns and find issues in the data frame.

In [224]:
# Program to print all the column names of the data_frame
list(data_frame.columns)


['Patient_ID',
 'Month_of_Birth',
 'Age',
 'Sex',
 'Occupation',
 'T_Stage',
 'N_Stage',
 '6th_Stage',
 'Differentiated',
 'Grade',
 'A_Stage',
 'Tumor_Size',
 'Estrogen_Status',
 'Progesterone_Status',
 'Regional_Node_Examined',
 'Reginol_Node_Positive',
 'Survival_Months',
 'Mortality_Status']

In [225]:
# Display data types and null counts to Find data formatting issues in the data_frame
data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 18 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Patient_ID              4024 non-null   object 
 1   Month_of_Birth          4024 non-null   int64  
 2   Age                     4015 non-null   float64
 3   Sex                     4020 non-null   object 
 4   Occupation              43 non-null     object 
 5   T_Stage                 4024 non-null   object 
 6   N_Stage                 4024 non-null   object 
 7   6th_Stage               4024 non-null   object 
 8   Differentiated          4024 non-null   object 
 9   Grade                   4024 non-null   int64  
 10  A_Stage                 4024 non-null   object 
 11  Tumor_Size              4021 non-null   float64
 12  Estrogen_Status         4024 non-null   object 
 13  Progesterone_Status     4024 non-null   object 
 14  Regional_Node_Examined  4023 non-null   

In [226]:
# print number of rows and columns in the data_frame
data_frame.shape

(4024, 18)

By using .describe() method, it is easy to identify the value errors in numerical columns. This code block is taken from code reuse session 1 and tutorial 1 documnent

In [227]:
# Display stats of numerical variables to Find value errors
data_frame.describe()

Month_of_Birth          Age        Grade   Tumor_Size  \
count     4024.000000  4015.000000  4024.000000  4021.000000   
mean         6.481362    54.107098     2.150596    30.419299   
std          3.475442    11.715528     0.638234    21.161080   
min          1.000000   -50.000000     1.000000   -75.000000   
25%          3.000000    47.000000     2.000000    16.000000   
50%          6.000000    54.000000     2.000000    25.000000   
75%         10.000000    61.000000     3.000000    38.000000   
max         12.000000   502.000000     4.000000   140.000000   

       Regional_Node_Examined  Reginol_Node_Positive  Survival_Months  
count             4023.000000            4024.000000      4024.000000  
mean                14.373602               4.158052        71.472167  
std                  8.129293               5.109331        25.361855  
min                  1.000000               1.000000         1.000000  
25%                  9.000000               1.000000        56.000000  
50%                 14.000000               2.000000        73.000000  
75%                 19.000000               5.000000        90.000000  
max                 61.000000              46.000000       760.000000

by using include=object, we can find the value errors for object  type (categorical) columns.

In [228]:
# Display stats of object type variables to Find value errors
data_frame.describe(include=object)


Patient_ID     Sex    Occupation T_Stage N_Stage 6th_Stage  \
count        4024    4020            43    4024    4024      4024   
unique       4024       2            40       4       3         5   
top         A4035  Female  House Person      T2      N1       IIA   
freq            1    4001             2    1786    2732      1305   

                   Differentiated   A_Stage Estrogen_Status  \
count                        4024      4024            4024   
unique                          4         2               2   
top     Moderately differentiated  Regional        Positive   
freq                         2351      3932            3755   

       Progesterone_Status Mortality_Status  
count                 4024             4024  
unique                   2                7  
top               Positive            Alive  
freq                  3326             3399

### **Mitigating Missing Data**

Below code block is reused from the code reuse session 1 to find the sum of missing values in the data frame

In [229]:
# Count of missing values per column
print("\nMissing values per column:")
print(data_frame.isnull().sum())


Missing values per column:
Patient_ID                   0
Month_of_Birth               0
Age                          9
Sex                          4
Occupation                3981
T_Stage                      0
N_Stage                      0
6th_Stage                    0
Differentiated               0
Grade                        0
A_Stage                      0
Tumor_Size                   3
Estrogen_Status              0
Progesterone_Status          0
Regional_Node_Examined       1
Reginol_Node_Positive        0
Survival_Months              0
Mortality_Status             0
dtype: int64


In [230]:
# Calculate and display percentage of missing data per column
missing_counts = data_frame.isna().sum()
missing_percent = (missing_counts / len(data_frame)) * 100
print("\nPercentage of missing data per column:")
print(missing_percent)


Percentage of missing data per column:
Patient_ID                 0.000000
Month_of_Birth             0.000000
Age                        0.223658
Sex                        0.099404
Occupation                98.931412
T_Stage                    0.000000
N_Stage                    0.000000
6th_Stage                  0.000000
Differentiated             0.000000
Grade                      0.000000
A_Stage                    0.000000
Tumor_Size                 0.074553
Estrogen_Status            0.000000
Progesterone_Status        0.000000
Regional_Node_Examined     0.024851
Reginol_Node_Positive      0.000000
Survival_Months            0.000000
Mortality_Status           0.000000
dtype: float64


According to the above stats, occupation has large sum of null values (3981) and this column is not important for the predictions

So, according to code reuse session 1, below code block is used to drop the unnessacery column from the data frame

In [231]:
# Dropping Occupation column cause it has too many null values it is irrrelavant feature
data_frame.drop(columns=['Occupation'], inplace=True)

Now the dataframe should have one less column (17)

In [232]:
data_frame.shape

(4024, 17)

Now dropping other irrelevant features which are not useful for predictions

In [233]:
# Dropping Patient ID, and Month_of_Birth columns
data_frame.drop(columns=['Month_of_Birth'], inplace=True)
data_frame.drop(columns=['Patient_ID'], inplace=True)

In [234]:
data_frame.shape

(4024, 15)

In [235]:
# Display data types and null counts
data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4024 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4015 non-null   float64
 1   Sex                     4020 non-null   object 
 2   T_Stage                 4024 non-null   object 
 3   N_Stage                 4024 non-null   object 
 4   6th_Stage               4024 non-null   object 
 5   Differentiated          4024 non-null   object 
 6   Grade                   4024 non-null   int64  
 7   A_Stage                 4024 non-null   object 
 8   Tumor_Size              4021 non-null   float64
 9   Estrogen_Status         4024 non-null   object 
 10  Progesterone_Status     4024 non-null   object 
 11  Regional_Node_Examined  4023 non-null   float64
 12  Reginol_Node_Positive   4024 non-null   int64  
 13  Survival_Months         4024 non-null   int64  
 14  Mortality_Status        4024 non-null   

In [236]:
# Display stats of the retained data frame
data_frame.describe()

Age        Grade   Tumor_Size  Regional_Node_Examined  \
count  4015.000000  4024.000000  4021.000000             4023.000000   
mean     54.107098     2.150596    30.419299               14.373602   
std      11.715528     0.638234    21.161080                8.129293   
min     -50.000000     1.000000   -75.000000                1.000000   
25%      47.000000     2.000000    16.000000                9.000000   
50%      54.000000     2.000000    25.000000               14.000000   
75%      61.000000     3.000000    38.000000               19.000000   
max     502.000000     4.000000   140.000000               61.000000   

       Reginol_Node_Positive  Survival_Months  
count            4024.000000      4024.000000  
mean                4.158052        71.472167  
std                 5.109331        25.361855  
min                 1.000000         1.000000  
25%                 1.000000        56.000000  
50%                 2.000000        73.000000  
75%                 5.000000        90.000000  
max                46.000000       760.000000

### **Imputing Numeric & categorical Features**

For numeric columns with limited missing values, it is possible to replace them by using th mean/median of each column.

This 'mean imputation' is suggessted in code reuse session 1 and below code blocks are from that session

For categorical columns with missing values, we will use 'mode imputation' (filling with the most frequent value in that column).

In [237]:
#Imputing the null values of Age with mean values
Mean_age =  data_frame['Age'].mean()
data_frame['Age'].fillna(Mean_age,inplace=True)

<ipython-input-237-ea68864691cd>:3: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [238]:
# fill missing Sex with the mode (most frequent)
Mode_Sex = data_frame['Sex'].mode()[0]
data_frame['Sex'].fillna(Mode_Sex,inplace=True)

<ipython-input-238-aa81748a2134>:3: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [239]:
#Imputing the null values of Tumor Size with mean values
Mean_Tumor_size = data_frame['Tumor_Size'].mean()
data_frame['Tumor_Size'].fillna(Mean_Tumor_size,inplace=True)

<ipython-input-239-dfcf02294460>:3: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [240]:
#Imputing the null values of Median_Reginol_Node_Examined with mean values
Median_Reginol_Node_Examined = data_frame['Regional_Node_Examined'].mean()
data_frame['Regional_Node_Examined'].fillna(Median_Reginol_Node_Examined,inplace=True)

<ipython-input-240-8226ef106054>:3: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.





In [241]:
#Re-check for any null values in dataset
data_frame.isnull().sum()

Age                       0
Sex                       0
T_Stage                   0
N_Stage                   0
6th_Stage                 0
Differentiated            0
Grade                     0
A_Stage                   0
Tumor_Size                0
Estrogen_Status           0
Progesterone_Status       0
Regional_Node_Examined    0
Reginol_Node_Positive     0
Survival_Months           0
Mortality_Status          0
dtype: int64

### **Identify Outliers and Extreme Values**

In this section, we identify and handle outliers in the dataset. We will use visualizations (histograms and boxplots) to find the extreme values and otliers, and then handle them accordingly.

Below code blocks are used from code reuse session 1 and tutorial 1 document

In [242]:
# Display box plot to discover outliers in Age feature
import plotly.express as px

Age_fig = px.box(data_frame, x='Age', title='Box Plot of Age')
Age_fig.show()

Below code block is reused from the code reuse session 1 to discover the outliers.

This method(function) finds and returns the outliers for a column using IQR

In [243]:
#Function to discover outliers using IQR
def find_outliers_IQR(data):
  q1=data.quantile(0.25)
  q3=data.quantile(0.75)
  IQR=q3-q1

  outliers = data[((data<(q1-1.5*IQR))|(data> (q3+1.5*IQR)))]

  return outliers

In [244]:
# Find and Print the outliers for Age
outliers = find_outliers_IQR(data_frame['Age'])

print("number of outliers: "+ str(len(outliers)))

print(outliers)

number of outliers: 4
142    180.0
212    -50.0
522    502.0
842     89.0
Name: Age, dtype: float64


This code block which is from code reuse session 1, iis used to remove the specific rows from daataframe which have identified as rows with outliers

In [245]:
# Drop records with outliers for Age
data_frame.drop(data_frame.index[[142, 212, 522,842]], inplace=True)

In [246]:
# Check number of rows after row removal for age - outliers
data_frame.shape

(4020, 15)

In [247]:
# Display box plot to discover outliers after removal in Age feature
import plotly.express as px

Age_fig = px.box(data_frame, x='Age', title='Box Plot of Age')
Age_fig.show()

Similary, In below code blocks, it finds outliers using IQR function and remove the rows accordingly to clean the dataset

In [248]:
# Display box plot to discover outliers in Survival_Months feature
import plotly.express as px

Survival_Months_fig = px.box(data_frame, x='Survival_Months', title='Box Plot of Survival_Months')
Survival_Months_fig.show()

In [249]:
# Find and Print the outliers for Survival month
outliers = find_outliers_IQR(data_frame['Survival_Months'])

print("number of outliers for Survival months: "+ str(len(outliers)))

outliers

number of outliers for Survival months: 19


374     760
414       3
679       4
721       3
896       4
921       1
928       3
1042      4
1156      2
1702      4
1709      3
1731      4
1751      2
1893      4
2230      2
2410      4
3089      4
3474      4
3803      4
Name: Survival_Months, dtype: int64

Only remove 374 because it seems to be an extreme value. Keep other rows without removing

In [250]:
# Drop the record with the outliers
data_frame.drop(data_frame.index[[374]], inplace=True)

In [251]:
# Check number of rows after row removal for outliers
data_frame.shape

(4019, 15)

In [252]:
# Display box plot to check outliers after removal in Survival_Months feature
import plotly.express as px

Survival_Months_fig = px.box(data_frame, x='Survival_Months', title='Box Plot of Survival_Months')
Survival_Months_fig.show()

In [253]:
# Display box plot to discover outliers in Tumor_Size feature
import plotly.express as px

Tumor_Size_fig = px.box(data_frame, x='Tumor_Size', title='Box Plot of Tumor_Size')
Tumor_Size_fig.show()

In [254]:
# Find and Print the outliers for Tumor Size using IQR function
outliers = find_outliers_IQR(data_frame['Tumor_Size'])

print("number of outliers: "+ str(len(outliers)))

outliers

number of outliers: 220


8       103.0
52      120.0
53       77.0
63       75.0
70      130.0
        ...  
3930    120.0
3965    140.0
3991     90.0
4009    100.0
4012     80.0
Name: Tumor_Size, Length: 220, dtype: float64

Since there are lots of outliers (nealry 220) for this column, we only select negative values as the outliers for tumer size column.

Because tumor size can not be negative. So, to identify the negative tumor size values, below function is used.

In [255]:
# Function to discover outliers with negative values
def find_outliers(data):

  outliers = data[((data<(0)))]

  return outliers

In [256]:
# Find and print outliers for Tumor size
outliers = find_outliers(data_frame['Tumor_Size'])

print("number of outliers: "+ str(len(outliers)))

outliers

number of outliers: 1


210   -75.0
Name: Tumor_Size, dtype: float64

In [257]:
# Drop the record with the outlier
data_frame.drop(data_frame.index[[210]], inplace=True, errors='ignore')

In [258]:
# Check number of rows after row removal for outliers
data_frame.shape

(4018, 15)

In [276]:
# Display box plot to check outliers aftter removing in Tumor_Size feature
import plotly.express as px

Tumor_Size_fig_new = px.box(data_frame, x='Tumor_Size', title='Box Plot of Tumor_Size')
Tumor_Size_fig_new.show()

Below code block is used from code reuse session 1 to find the distinct values in Mortality status feature.

In [260]:
# Analys different values in Mortality Status column
data_frame['Mortality_Status'].unique()

array(['Alive', 'Dead', 'ALIVE', 'DEAD', 'ALive', 'alive', 'dead'],
      dtype=object)

In [261]:
# Mapping values in catergorical column, Mortality Status
# Alive -> 0
# Dead -> 1
data_frame['Mortality_Status'] = data_frame['Mortality_Status'].str.lower().map({'alive': 0, 'dead': 1})

In [262]:
# Check unique values in Mortality_Status column after updating
data_frame['Mortality_Status'].unique()

array([0, 1])

In [263]:
data_frame.head(10)

Age     Sex T_Stage N_Stage 6th_Stage             Differentiated  Grade  \
0  68.0  Female      T1      N1       IIA      Poorly differentiated      3   
1  50.0  Female      T2      N2      IIIA  Moderately differentiated      2   
2  58.0  Female      T3      N3      IIIC  Moderately differentiated      2   
3  58.0  Female      T1      N1       IIA      Poorly differentiated      3   
4  47.0  Female      T2      N1       IIB      Poorly differentiated      3   
5  51.0  Female      T1      N1       IIA  Moderately differentiated      2   
6  51.0  Female      T1      N1       IIA        Well differentiated      1   
7  40.0  Female      T2      N1       IIB  Moderately differentiated      2   
8  40.0  Female      T4      N3      IIIC      Poorly differentiated      3   
9  69.0  Female      T4      N3      IIIC        Well differentiated      1   

    A_Stage  Tumor_Size Estrogen_Status Progesterone_Status  \
0  Regional         4.0        Positive            Positive   
1  Regional        35.0        Positive            Positive   
2  Regional        63.0        Positive            Positive   
3  Regional        18.0        Positive            Positive   
4  Regional        41.0        Positive            Positive   
5  Regional        20.0        Positive            Positive   
6  Regional         8.0        Positive            Positive   
7  Regional        30.0        Positive            Positive   
8  Regional       103.0        Positive            Positive   
9   Distant        32.0        Positive            Positive   

   Regional_Node_Examined  Reginol_Node_Positive  Survival_Months  \
0                    24.0                      1               60   
1                    14.0                      5               62   
2                    14.0                      7               75   
3                     2.0                      1               84   
4                     3.0                      1               50   
5                    18.0                      2               89   
6                    11.0                      1               54   
7                     9.0                      1               14   
8                    20.0                     18               70   
9                    21.0                     12               92   

   Mortality_Status  
0                 0  
1                 0  
2                 0  
3                 0  
4                 0  
5                 0  
6                 0  
7                 1  
8                 0  
9                 0

### **Transforming categorical values to ordinal**

In sex feature, it contains inconsistence values. (both string and numeric). So it should be replaced into numeric values as 1 and 0 (female and male)

for that the same .map code function from code reuse session 1 is used

In [264]:
# Check unique values in sex column
data_frame['Sex'].unique()

array(['Female', '1'], dtype=object)

In [265]:
# '1' for Male  &
data_frame['Sex'].replace({'Female': 0, '1': 1}, inplace=True)

<ipython-input-265-8ecfaf5e611c>:2: FutureWarning:

A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.



<ipython-input-265-8ecfaf5e611c>:2: FutureWarning:

Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`



In [266]:
# Check unique values in sex column
data_frame['Sex'].unique()

array([0, 1])

In [267]:
data_frame.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4018 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4018 non-null   float64
 1   Sex                     4018 non-null   int64  
 2   T_Stage                 4018 non-null   object 
 3   N_Stage                 4018 non-null   object 
 4   6th_Stage               4018 non-null   object 
 5   Differentiated          4018 non-null   object 
 6   Grade                   4018 non-null   int64  
 7   A_Stage                 4018 non-null   object 
 8   Tumor_Size              4018 non-null   float64
 9   Estrogen_Status         4018 non-null   object 
 10  Progesterone_Status     4018 non-null   object 
 11  Regional_Node_Examined  4018 non-null   float64
 12  Reginol_Node_Positive   4018 non-null   int64  
 13  Survival_Months         4018 non-null   int64  
 14  Mortality_Status        4018 non-null   int64

In [268]:
# Check unique values in Differentiated column
data_frame['Differentiated'].unique()

array(['Poorly differentiated', 'Moderately differentiated',
       'Well differentiated', 'Undifferentiated'], dtype=object)

In [269]:
# Import preprocessing
from sklearn import preprocessing

# Initialize the LabelEncoder to encode ordinal columns
label_encoder = preprocessing.LabelEncoder()

# Fit and transform the 'Estrogen_Status' column
data_frame['Estrogen_Status']= label_encoder.fit_transform(data_frame['Estrogen_Status'])

# Fit and transform the 'Progesterone_Status' column
data_frame['Progesterone_Status']= label_encoder.fit_transform(data_frame['Progesterone_Status'])
print(data_frame.info())


<class 'pandas.core.frame.DataFrame'>
Index: 4018 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4018 non-null   float64
 1   Sex                     4018 non-null   int64  
 2   T_Stage                 4018 non-null   object 
 3   N_Stage                 4018 non-null   object 
 4   6th_Stage               4018 non-null   object 
 5   Differentiated          4018 non-null   object 
 6   Grade                   4018 non-null   int64  
 7   A_Stage                 4018 non-null   object 
 8   Tumor_Size              4018 non-null   float64
 9   Estrogen_Status         4018 non-null   int64  
 10  Progesterone_Status     4018 non-null   int64  
 11  Regional_Node_Examined  4018 non-null   float64
 12  Reginol_Node_Positive   4018 non-null   int64  
 13  Survival_Months         4018 non-null   int64  
 14  Mortality_Status        4018 non-null   int64

In [270]:
#Fit and transform all ordinal columns
data_frame['T_Stage']= label_encoder.fit_transform(data_frame['T_Stage'])
data_frame['N_Stage']= label_encoder.fit_transform(data_frame['N_Stage'])
data_frame['6th_Stage']= label_encoder.fit_transform(data_frame['6th_Stage'])
data_frame['Differentiated']= label_encoder.fit_transform(data_frame['Differentiated'])
data_frame['A_Stage']= label_encoder.fit_transform(data_frame['A_Stage'])

print(data_frame.info())

<class 'pandas.core.frame.DataFrame'>
Index: 4018 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Age                     4018 non-null   float64
 1   Sex                     4018 non-null   int64  
 2   T_Stage                 4018 non-null   int64  
 3   N_Stage                 4018 non-null   int64  
 4   6th_Stage               4018 non-null   int64  
 5   Differentiated          4018 non-null   int64  
 6   Grade                   4018 non-null   int64  
 7   A_Stage                 4018 non-null   int64  
 8   Tumor_Size              4018 non-null   float64
 9   Estrogen_Status         4018 non-null   int64  
 10  Progesterone_Status     4018 non-null   int64  
 11  Regional_Node_Examined  4018 non-null   float64
 12  Reginol_Node_Positive   4018 non-null   int64  
 13  Survival_Months         4018 non-null   int64  
 14  Mortality_Status        4018 non-null   int64

 There are some float value variables but decimal precision of these variables’ values is not of a high importance in real-world applications.


 Therefore, below code cell from code reuse session 1 is used to convert the float variables into integer data type



In [271]:
# Converted the 3 float data types into int type
data_frame = data_frame.astype({'Age':'int','Tumor_Size':'int','Regional_Node_Examined':'int'})

In [272]:
# Display data types
data_frame.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4018 entries, 0 to 4023
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   Age                     4018 non-null   int64
 1   Sex                     4018 non-null   int64
 2   T_Stage                 4018 non-null   int64
 3   N_Stage                 4018 non-null   int64
 4   6th_Stage               4018 non-null   int64
 5   Differentiated          4018 non-null   int64
 6   Grade                   4018 non-null   int64
 7   A_Stage                 4018 non-null   int64
 8   Tumor_Size              4018 non-null   int64
 9   Estrogen_Status         4018 non-null   int64
 10  Progesterone_Status     4018 non-null   int64
 11  Regional_Node_Examined  4018 non-null   int64
 12  Reginol_Node_Positive   4018 non-null   int64
 13  Survival_Months         4018 non-null   int64
 14  Mortality_Status        4018 non-null   int64
dtypes: int64(15)
memory usage:

In [273]:
# Display stats of the retained data frame
data_frame.describe()

Age          Sex      T_Stage      N_Stage    6th_Stage  \
count  4018.000000  4018.000000  4018.000000  4018.000000  4018.000000   
mean     53.986560     0.004729     0.784470     0.437531     1.320060   
std       8.953942     0.068611     0.765407     0.692580     1.265226   
min      30.000000     0.000000     0.000000     0.000000     0.000000   
25%      47.000000     0.000000     0.000000     0.000000     0.000000   
50%      54.000000     0.000000     1.000000     0.000000     1.000000   
75%      61.000000     0.000000     1.000000     1.000000     2.000000   
max      69.000000     1.000000     3.000000     2.000000     4.000000   

       Differentiated        Grade      A_Stage   Tumor_Size  Estrogen_Status  \
count     4018.000000  4018.000000  4018.000000  4018.000000      4018.000000   
mean         0.689895     2.151319     0.977352    30.408163         0.933051   
std          1.015637     0.638176     0.148797    21.135071         0.249964   
min          0.000000     1.000000     0.000000   -75.000000         0.000000   
25%          0.000000     2.000000     1.000000    16.000000         1.000000   
50%          0.000000     2.000000     1.000000    25.000000         1.000000   
75%          1.000000     3.000000     1.000000    38.000000         1.000000   
max          3.000000     4.000000     1.000000   140.000000         1.000000   

       Progesterone_Status  Regional_Node_Examined  Reginol_Node_Positive  \
count          4018.000000             4018.000000            4018.000000   
mean              0.826531               14.366103               4.146590   
std               0.378700                8.129345               5.091756   
min               0.000000                1.000000               1.000000   
25%               1.000000                9.000000               1.000000   
50%               1.000000               14.000000               2.000000   
75%               1.000000               19.000000               5.000000   
max               1.000000               61.000000              46.000000   

       Survival_Months  Mortality_Status  
count      4018.000000       4018.000000  
mean         71.489796          0.152812  
std          25.349448          0.359851  
min           1.000000          0.000000  
25%          56.000000          0.000000  
50%          73.000000          0.000000  
75%          90.000000          0.000000  
max         760.000000          1.000000

In [274]:
# Prepare classification dataset
data_frame_classification =data_frame.drop(columns=['Survival_Months']).copy()

# Prepare regression dataset
data_frame_regression = data_frame.drop(columns=['Mortality_Status']).copy()

In [275]:
# Save the cleaned and prepared datasets to CSV files
data_frame_classification.to_csv("cleaned_classification_data.csv", index=False)
data_frame_regression.to_csv("cleaned_regression_data.csv", index=False)